### ISSUE TITLE
Warning if packages --upgrade cannot happen for the cutrent nixOS version / channel

### ISSUE THREAD

**Issue by adingbatponder** [NixOS/nixpkgs#355018]
`sudo nixos-rebuild switch --upgrade`

does not yield a message warning that no upgrade is necessarily made to all packages when it is run, when run at a time when is the case that the current OS version, e.g. 23.11, is past the end of support date and so no packages are upgraded any more in the store for that OS version. A useful message could say that some of the packages in the configuration.nix may require a nixOS upgrade first in order to be upgraded at all. A list of affected passages would be optimal.

The issue is important because at boot the warning that the nixOS is out of date shoots past and is not legible (and nixOS is so stable that rebooting may not take place ofte ) and the consequences for the --upgrade flag are not clear.

![Screenshot_20241110-125151~2](https://github.com/user-attachments/assets/f30b3e57-bbc6-4389-8b83-d1e83fb572e0)

Thanks.

**Comment by misuzu** [comment]
The NixOS 23.11 is EOL and doesn't receive updates. To receive updates you have to upgrade to NixOS 24.05 or NixOS unstable. Here's a [manual entry](https://nixos.org/manual/nixos/stable/#sec-upgrading) for upgrading NixOS to the next stable release.

tl;dr:
```
sudo nix-channel --add https://channels.nixos.org/nixos-24.05 nixos
sudo nix-channel --update nixos
sudo nixos-rebuild boot
sudo reboot
```

**Comment by LunNova** [comment]
misuzu, I think you missed OP's concern. It's not that they don't know how to upgrade, it's that it's easy to miss the outdated release warning that shows up at boot and we should consider adding the same warning to nixos-rebuild.

**Comment by misuzu** [comment]
There's nothing NixOS-specific that makes that warning, it's a feature of systemd. It just compares [`SUPPORT_END` from `/etc/os-release`](https://github.com/NixOS/nixpkgs/blob/205fd4226592cc83fd4c0885a3e4c9c400efabb5/nixos/modules/misc/version.nix#L30) with current date and shows that warning.

The current nixos-rebuild is already a mess and it's better to not make it worse. The [nixos-rebuild-ng](https://github.com/NixOS/nixpkgs/pull/354029) could probably add a similar logic.

Though if this feature implemented like that it could lead to a confusing situation, e.g. a system runs outdated release and user added a fresh channel but the `nixos-rebuild` would still spit out the warning.
